File: kuikly/library/shared/src/commonMain/kotlin/com/techskillplanet/phonics/controls/TspTag.kt
package com.techskillplanet.phonics.controls

import androidx.compose.runtime.Composable
import com.tencent.kuikly.compose.foundation.background
import com.tencent.kuikly.compose.foundation.border
import com.tencent.kuikly.compose.foundation.clickable
import com.tencent.kuikly.compose.foundation.layout.Arrangement
import com.tencent.kuikly.compose.foundation.layout.Row
import com.tencent.kuikly.compose.foundation.layout.padding
import com.tencent.kuikly.compose.foundation.layout.size
import com.tencent.kuikly.compose.foundation.shape.RoundedCornerShape
import com.tencent.kuikly.compose.material3.Text
import com.tencent.kuikly.compose.ui.Alignment
import com.tencent.kuikly.compose.ui.Modifier
import com.tencent.kuikly.compose.ui.graphics.Color
import com.tencent.kuikly.compose.ui.text.font.FontWeight
import com.tencent.kuikly.compose.ui.unit.dp
import com.tencent.kuikly.compose.ui.unit.sp
import com.techskillplanet.phonics.theme.PhonicsColors

@Composable
fun TspTag(
    theme: PhonicsColors,
    modifier: Modifier = Modifier,
    text: String = "",
    closable: Boolean = false,
    selected: Boolean = false,
    disabled: Boolean = false,
    variant: String = "default",
    onClose: () -> Unit = {},
    onTap: () -> Unit = {},
) {
    val (bg, fg, border) = tagColors(theme, variant)
    Row(
        modifier = modifier
            .background(Color(bg), RoundedCornerShape(999.dp))
            .border(
                if (selected) 2.dp else 1.dp,
                Color(if (selected) theme.selectedBorder else border),
                RoundedCornerShape(999.dp),
            )
            .clickable(enabled = !disabled) { onTap() }
            .padding(horizontal = 12.dp, vertical = 6.dp),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(4.dp),
    ) {
        Text(
            text = text,
            color = Color(if (disabled) theme.textTertiary else fg),
            fontWeight = FontWeight.Bold,
            fontSize = 13.sp,
        )
        if (closable) {
            Text(
                text = "×",
                color = Color(if (disabled) theme.textTertiary else fg),
                fontWeight = FontWeight.Bold,
                fontSize = 16.sp,
                modifier = Modifier
                    .size(20.dp)
                    .clickable(enabled = !disabled) { onClose() },
            )
        }
    }
}

private fun tagColors(theme: PhonicsColors, variant: String): Triple<Long, Long, Long> {
    return when (variant) {
        "primary" -> Triple(theme.brandSubtle, theme.brandPrimary, 0x00000000)
        "success" -> Triple(theme.successSubtle, theme.success, 0x00000000)
        "warning" -> Triple(theme.activeFill, theme.textPrimary, 0x00000000)
        "danger" -> Triple(theme.dangerSoft, theme.danger, 0x00000000)
        else -> Triple(theme.surfaceRaised, theme.textPrimary, theme.borderDefault)
    }
}

File: react-native/library/src/starPlanet/components/TspTag.js
import React from 'react';
import { Pressable, Text, View, StyleSheet } from 'react-native';
import { styles as shared, withTheme } from '../utils/shared';

function tagColors(t, variant) {
  switch (variant) {
    case 'primary':
      return { bg: t.brandSubtle, color: t.brandPrimary, border: 'transparent' };
    case 'success':
      return { bg: t.successSubtle, color: t.success, border: 'transparent' };
    case 'warning':
      return { bg: t.activeFill, color: t.textPrimary, border: 'transparent' };
    case 'danger':
      return { bg: '#FFE8EB', color: t.danger, border: 'transparent' };
    default:
      return { bg: t.surfaceRaised, color: t.textPrimary, border: t.borderDefault };
  }
}

/** Closable / selectable tag (Chip remains for filter chips). */
export function TspTag({
  text = '',
  closable = false,
  selected = false,
  disabled = false,
  variant = 'default',
  theme,
  onClose,
  onTap,
}) {
  const t = withTheme(theme);
  const colors = tagColors(t, variant);

  const Root = onTap ? Pressable : View;
  const rootProps = onTap
    ? {
        accessibilityRole: 'button',
        accessibilityState: { selected, disabled },
        disabled,
        onPress: () => {
          if (!disabled) onTap?.();
        },
      }
    : { accessibilityRole: 'text' };

  return (
    <Root
      accessibilityLabel={text || 'Tag'}
      {...rootProps}
      style={[
        local.root,
        {
          backgroundColor: colors.bg,
          borderColor: selected ? t.selectedBorder : colors.border,
          borderWidth: selected ? 2 : 1,
        },
        disabled && shared.disabled,
      ]}
    >
      <Text style={[local.text, { color: colors.color }]} numberOfLines={1}>
        {text}
      </Text>
      {closable ? (
        <Pressable
          accessibilityRole="button"
          accessibilityLabel="Remove"
          disabled={disabled}
          hitSlop={6}
          onPress={(e) => {
            e?.stopPropagation?.();
            if (!disabled) onClose?.();
          }}
          style={local.close}
        >
          <Text style={[local.closeGlyph, { color: colors.color }]}>×</Text>
        </Pressable>
      ) : null}
    </Root>
  );
}

const local = StyleSheet.create({
  root: {
    flexDirection: 'row',
    alignItems: 'center',
    alignSelf: 'flex-start',
    gap: 4,
    maxWidth: '100%',
    borderRadius: 999,
    paddingHorizontal: 12,
    paddingVertical: 6,
  },
  text: { fontSize: 13, fontWeight: '700', flexShrink: 1 },
  close: {
    width: 20,
    height: 20,
    borderRadius: 10,
    alignItems: 'center',
    justifyContent: 'center',
  },
  closeGlyph: { fontSize: 16, fontWeight: '700', lineHeight: 18 },
});
